core/index/index_writer.cpp

```diff
@@ -804,83 +778,59 @@ void IndexWriter::Transaction::Reset() noexcept {
   }
 
   // rollback modification queries
-  std::for_each(std::begin(ctx->modification_queries_) +
-                  ctx->uncomitted_modification_queries_,
-                std::end(ctx->modification_queries_),
-                [](ModificationContext& ctx) noexcept {
-                  // mark invalid
-                  ctx.filter = nullptr;
-                });
+  std::for_each(
+    ctx->modification_queries_.begin() + ctx->uncommitted_modification_queries_,
+    ctx->modification_queries_.end(), [](ModificationContext& ctx) noexcept {
+      ctx.filter = nullptr;  // mark invalid
+    });
 
   auto& flushed_update_contexts = ctx->flushed_update_contexts_;
 
-  // find and mask/truncate uncomitted tail
-  for (size_t i = 0, count = ctx->flushed_.size(), flushed_docs_count = 0;
-       i < count; ++i) {
-    auto& segment = ctx->flushed_[i];
-    auto flushed_docs_start = flushed_docs_count;
-
-    // sum of all previous SegmentMeta::docs_count
-    // including this meta
-    flushed_docs_count += segment.meta.docs_count;
-
-    if (flushed_docs_count <=
-        ctx->uncomitted_doc_id_begin_ - doc_limits::min()) {
-      continue;  // all documents in this this index_meta have been committed
-    }
-
-    auto docs_mask_tail_doc_id =
-      ctx->uncomitted_doc_id_begin_ - flushed_docs_start;
-
-    IRS_ASSERT(docs_mask_tail_doc_id <= segment.meta.live_docs_count);
-    IRS_ASSERT(docs_mask_tail_doc_id <= std::numeric_limits<doc_id_t>::max());
-    segment.docs_mask_tail_doc_id = doc_id_t(docs_mask_tail_doc_id);
-
-    if (docs_mask_tail_doc_id - doc_limits::min() >= segment.meta.docs_count) {
-      ctx->flushed_.resize(i);  // truncate including current empty meta
+  // find and mask/truncate uncommitted tail
+  const auto end = ctx->flushed_.end();
+  auto it = std::find_if(ctx->flushed_.begin(), end, [&](const auto& flushed) {
+    return ctx->uncommitted_docs_ < flushed.GetDocsEnd();
+  });
+  if (it != end) {
+    if (it->SetCommitted(ctx->uncommitted_docs_)) {
+      const auto docs_end = it->GetDocsEnd();
+      ctx->flushed_.erase(it + 1, end);
+      ctx->flushed_update_contexts_.resize(docs_end);
+      ctx->uncommitted_docs_ = docs_end;
     } else {
-      ctx->flushed_.resize(i + 1);  // truncate starting from subsequent meta
+      const auto docs_begin = it->GetDocsBegin();
```

Comment:
SetCommitted returns false only if  `docs_begin` == 'committed' which is `ctx->uncommitted_docs_`
So why not just do in else:
 ctx->flushed_.erase(it, end);
 ctx->flushed_update_contexts_.resize(ctx->uncommitted_docs_);

Reply:
Ok